README.rst
verkkomaksut
------------

|build status|_

.. |build status| image:: https://secure.travis-ci.org/fastmonkeys/python-verkkomaksut.png?branch=master
   :alt: Build Status
.. _build status: http://travis-ci.org/fastmonkeys/python-verkkomaksut

Python wrapper for the JSON API of Suomen Verkkomaksut.

Installation
------------

::

    $ pip install verkkomaksut

Resources
---------

- `Issue Tracker <http://github.com/fastmonkeys/python-verkkomaksut/issues>`_
- `Code <http://github.com/fastmonkeys/python-verkkomaksut>`_
- `Development Version
  <http://github.com/fastmonkeys/python-verkkomaksut/zipball/master#egg=verkkomaksut-dev>`_


CHANGES.rst
Changelog
---------

Here you can see the full list of changes between each python-verkkomaksut
release.

0.2.0 (June 6, 2013)
^^^^^^^^^^^^^^^^^^^^

- Initial public release.


test_verkkomaksut.py
# -*- coding: utf-8 -*-
"""
    test_verkkomaksut
    ~~~~~~~~~~~~~~~~~

    Tests for `verkkomaksut` module.

    :copyright: (c) 2012 by Janne Vanhala.
    :license: BSD, see LICENSE for more details.
"""
import requests
from flexmock import flexmock
from pytest import raises
from verkkomaksut import (
    Client,
    Contact,
    Payment,
    Product,
    VerkkomaksutException
)


class TestVerkkomaksutException(object):
    def setup_method(self, method):
        self.exception = VerkkomaksutException('code', 'message')

    def test_is_an_exception(self):
        assert isinstance(self.exception, Exception)

    def test_sets_error_code(self):
        assert self.exception.code == 'code'

    def test_sets_error_message(self):
        assert self.exception.message == 'message'


class TestContactWithMinimumParameters(object):
    def setup_method(self, method):
        self.contact = Contact(
            first_name='Matti',
            last_name='Meikäläinen',
            email='matti.meikalainen@gmail.com',
            street='Esimerkkikatu 123',
            postal_code='01234',
            postal_office='Helsinki',
            country='FI'
        )

    def test_sets_first_name(self):
        assert self.contact.first_name == 'Matti'

    def test_sets_last_name(self):
        assert self.contact.last_name == 'Meikäläinen'

    def test_sets_email(self):
        assert self.contact.email == 'matti.meikalainen@gmail.com'

    def test_sets_street(self):
        assert self.contact.street == 'Esimerkkikatu 123'

    def test_sets_postal_code(self):
        assert self.contact.postal_code == '01234'

    def test_sets_postal_office(self):
        assert self.contact.postal_office == 'Helsinki'

    def test_sets_country(self):
        assert self.contact.country == 'FI'

    def test_telephone_is_none_by_default(self):
        assert self.contact.telephone is None

    def test_mobile_is_none_by_default(self):
        assert self.contact.mobile is None

    def test_company_name_is_none_by_default(self):
        assert self.contact.company_name is None


class TestContactWithAllParameters(object):
    def setup_method(self, method):
        self.contact = Contact(
            first_name='Matti',
            last_name='Meikäläinen',
            email='matti.meikalainen@gmail.com',
            street='Esimerkkikatu 123',
            postal_code='01234',
            postal_office='Helsinki',
            country='FI',
            telephone='020123456',
            mobile='050123456',
            company_name='Yhtiö'
        )

    def test_sets_telephone(self):
        assert self.contact.telephone == '020123456'

    def test_sets_mobile(self):
        assert self.contact.mobile == '050123456'

    def test_sets_company_name(self):
        assert self.contact.company_name == 'Yhtiö'

    def test_json(self):
        assert self.contact.json == {
            'telephone': '020123456',
            'mobile': '050123456',
            'email': 'matti.meikalainen@gmail.com',
            'firstName': 'Matti',
            'lastName': 'Meikäläinen',
            'companyName': 'Yhtiö',
            'address': {
                'street': 'Esimerkkikatu 123',
                'postalCode': '01234',
                'postalOffice': 'Helsinki',
                'country': 'FI',
            },
        }


class TestProductWithMinimumParameters(object):
    def setup_method(self, method):
        self.product = Product(
            title='Esimerkkituote',
            price='19.90',
            vat='23.00'
        )

    def test_sets_title(self):
        assert self.product.title == 'Esimerkkituote'

    def test_sets_price(self):
        assert self.product.price == '19.90'

    def test_sets_vat(self):
        assert self.product.vat == '23.00'

    def test_amount_is_one_by_default(self):
        assert self.product.amount == 1

    def test_code_is_none_by_default(self):
        assert self.product.code is None

    def test_discount_is_zero_by_default(self):
        assert self.product.discount == 0

    def test_type_is_normal_by_default(self):
        assert self.product.type == Product.TYPE_NORMAL

    def test_type_can_be_normal(self):
        self.product.type = Product.TYPE_NORMAL
        assert self.product.type == Product.TYPE_NORMAL

    def test_type_can_be_postage(self):
        self.product.type = Product.TYPE_POSTAGE
        assert self.product.type == Product.TYPE_POSTAGE

    def test_type_can_be_processing(self):
        self.product.type = Product.TYPE_PROCESSING
        assert self.product.type == Product.TYPE_PROCESSING

    def test_raises_value_error_on_unsupported_type(self):
        with raises(ValueError) as exc_info:
            self.product.type = 0
        assert exc_info.value.message == 'Given product type not supported: 0'


class TestProductWithAllParameters(object):
    def setup_method(self, method):
        self.product = Product(
            title='Esimerkkituote',
            price='19.90',
            vat=23,
            amount=2,
            code='XX-123',
            discount=50,
            type=Product.TYPE_POSTAGE
        )

    def test_sets_amount(self):
        assert self.product.amount == 2

    def test_sets_code(self):
        assert self.product.code == 'XX-123'

    def test_sets_discount(self):
        assert self.product.discount == 50

    def test_sets_type(self):
        assert self.product.type == Product.TYPE_POSTAGE

    def test_json(self):
        assert self.product.json == {
            'title': 'Esimerkkituote',
            'code': 'XX-123',
            'amount': 2,
            'price': '19.90',
            'vat': 23,
            'discount': 50,
            'type': 2
        }


class TestPaymentWithMinimumParameters(object):
    def setup_method(self, method):
        self.contact = Contact(
            first_name='Matti',
            last_name='Meikäläinen',
            email='matti.meikalainen@gmail.com',
            street='Esimerkkikatu 123',
            postal_code='01234',
            postal_office='Helsinki',
            country='FI'
        )
        self.payment = Payment(
            order_number='12345678',
            contact=self.contact,
            success_url='https://www.esimerkkikauppa.fi/sv/success',
            failure_url='https://www.esimerkkikauppa.fi/sv/failure',
            notification_url='https://www.esimerkkikauppa.fi/sv/success'
        )

    def test_sets_order_number(self):
        assert self.payment.order_number == '12345678'

    def test_sets_contact(self):
        assert self.payment.contact is self.contact

    def test_sets_success_url(self):
        assert self.payment.success_url == \
            'https://www.esimerkkikauppa.fi/sv/success'

    def test_sets_failure_url(self):
        assert self.payment.failure_url == \
            'https://www.esimerkkikauppa.fi/sv/failure'

    def test_sets_notification_url(self):
        assert self.payment.notification_url == \
            'https://www.esimerkkikauppa.fi/sv/success'

    def test_product_list_is_empty(self):
        assert len(self.payment.products) == 0

    def test_reference_number_is_none_by_default(self):
        assert self.payment.reference_number is None

    def test_description_is_none_by_default(self):
        assert self.payment.description is None

    def test_currency_is_euro_by_default(self):
        assert self.payment.currency == 'EUR'

    def test_locale_is_finnish_by_default(self):
        assert self.payment.locale == 'fi_FI'

    def test_vat_is_included_by_default(self):
        assert self.payment.include_vat

    def test_pending_url_is_none_by_default(self):
        assert self.payment.pending_url is None

    def test_unsupported_currency_raises_value_error(self):
        with raises(ValueError) as exc_info:
            self.payment.currency = 'USD'
        assert exc_info.value.message == \
            'Currently EUR is the only supported currency.'

    def test_locale_can_be_finnish(self):
        self.payment.locale = 'fi_FI'
        assert self.payment.locale == 'fi_FI'

    def test_locale_can_be_swedish(self):
        self.payment.locale = 'sv_SE'
        assert self.payment.locale == 'sv_SE'

    def test_locale_can_be_english(self):
        self.payment.locale = 'en_US'
        assert self.payment.locale == 'en_US'

    def test_unsupported_locale_raises_value_error(self):
        with raises(ValueError) as exc_info:
            self.payment.locale = 'de_DE'
        assert exc_info.value.message == \
            "Given locale is not supported: 'de_DE'"

    def test_json(self):
        assert self.payment.json == {
            'orderNumber': '12345678',
            'referenceNumber': None,
            'description': None,
            'currency': 'EUR',
            'locale': 'fi_FI',
            'urlSet': {
                'success': 'https://www.esimerkkikauppa.fi/sv/success',
                'pending': None,
                'failure': 'https://www.esimerkkikauppa.fi/sv/failure',
                'notification': 'https://www.esimerkkikauppa.fi/sv/success'
            },
            'orderDetails': {
                'includeVat': '1',
                'contact': self.contact.json,
                'products': []
            }
        }


class TestPaymentWithAllParameters(object):
    def setup_method(self, method):
        self.contact = Contact(
            first_name='Matti',
            last_name='Meikäläinen',
            email='matti.meikalainen@gmail.com',
            street='Esimerkkikatu 123',
            postal_code='01234',
            postal_office='Helsinki',
            country='FI'
        )
        self.payment = Payment(
            order_number='12345678',
            contact=self.contact,
            reference_number='12345-12345-12345',
            description='Esimerkkimaksun kuvaus',
            currency='EUR',
            locale='sv_SE',
            include_vat=False,
            success_url='https://www.esimerkkikauppa.fi/sv/success',
            pending_url='https://www.esimerkkikauppa.fi/sv/pending',
            failure_url='https://www.esimerkkikauppa.fi/sv/failure',
            notification_url='https://www.esimerkkikauppa.fi/sv/success'
        )
        self.payment.products.append(
            Product(
                title='Esimerkkituote',
                price='19.90',
                vat='23.00'
            )
        )

    def test_sets_reference_number(self):
        assert self.payment.reference_number == '12345-12345-12345'

    def test_sets_description(self):
        assert self.payment.description == 'Esimerkkimaksun kuvaus'

    def test_sets_currency(self):
        assert self.payment.currency == 'EUR'

    def test_sets_locale(self):
        assert self.payment.locale == 'sv_SE'

    def test_sets_include_vat(self):
        assert not self.payment.include_vat

    def test_sets_pending_url(self):
        assert self.payment.pending_url == \
            'https://www.esimerkkikauppa.fi/sv/pending'

    def test_json(self):
        assert self.payment.json == {
            'orderNumber': '12345678',
            'referenceNumber': '12345-12345-12345',
            'description': 'Esimerkkimaksun kuvaus',
            'currency': 'EUR',
            'locale': 'sv_SE',
            'urlSet': {
                'success': 'https://www.esimerkkikauppa.fi/sv/success',
                'failure': 'https://www.esimerkkikauppa.fi/sv/failure',
                'pending': 'https://www.esimerkkikauppa.fi/sv/pending',
                'notification': 'https://www.esimerkkikauppa.fi/sv/success'
            },
            'orderDetails': {
                'includeVat': '0',
                'contact': self.contact.json,
                'products': [self.payment.products[0].json]
            }
        }


class MockPayment(object):
    def __init__(self, order_number='12345'):
        self.order_number = order_number

    @property
    def json(self):
        return {
            'orderNumber': self.order_number
        }


class TestClient(object):
    def test_defaults_to_merchant_test_account(self):
        client = Client()
        assert client.session.auth == \
            ('13466', '6pKF4jkv97zmqBJ3ZL8gUw5DfT2NMQ')

    def test_custom_merchant_credentials(self):
        client = Client(merchant_id='12345', merchant_secret='secret')
        assert client.session.auth == ('12345', 'secret')

    def test_request_content_type_is_json(self):
        client = Client()
        assert client.session.headers['Content-Type'] == 'application/json'

    def test_request_accepts_json(self):
        client = Client()
        assert client.session.headers['Accept'] == 'application/json'

    def test_request_defines_api_version(self):
        client = Client()
        assert client.session.headers['X-Verkkomaksut-Api-Version'] == '1'

    def test_successful_payment_creation(self):
        response = requests.Response()
        response._content = """{
  "orderNumber": "12345",
  "token": "secret_token",
  "url": "https://payment.verkkomaksut.fi/payment/load/token/secret_token"
}"""
        response.status_code = requests.codes.created

        client = Client()
        flexmock(client.session)
        client.session \
            .should_receive('post') \
            .with_args(
                'https://payment.verkkomaksut.fi/api-payment/create',
                data='{"orderNumber": "12345"}',
            ) \
            .and_return(response)

        payment = MockPayment(order_number='12345')
        assert client.create_payment(payment) == {
            'order_number': '12345',
            'token': 'secret_token',
            'url': 'https://payment.verkkomaksut.fi/payment/load/token/'
                   'secret_token'
        }

    def test_payment_creation_failure(self):
        response = requests.Response()
        response._content = """{
  "errorCode": "invalid-order-number",
  "errorMessage": "Missing or invalid order number"
}"""
        response.status_code = requests.codes.bad_request

        client = Client()
        flexmock(client.session)
        client.session \
            .should_receive('post') \
            .with_args(
                'https://payment.verkkomaksut.fi/api-payment/create',
                data='{"orderNumber": "12345"}',
            ) \
            .and_return(response)

        payment = MockPayment()
        with raises(VerkkomaksutException) as exc_info:
            client.create_payment(payment)

        assert exc_info.value.code == 'invalid-order-number'
        assert exc_info.value.message == 'Missing or invalid order number'

    def test_calculate_payment_receipt_hash(self):
        client = Client()
        hash_ = client._calculate_payment_receipt_hash(
            '15153',
            '1176557554',
            '012345ABCDE',
            '1'
        )
        assert hash_ == '555E0C0DE304938AACA5D594DB72F315'

    def test_validate_payment_receipt_parameters_ok(self):
        client = Client()
        flexmock(client) \
            .should_receive('_calculate_payment_receipt_hash') \
            .with_args('order_number', 'timestamp', 'paid', 'method') \
            .and_return('authcode')
        assert client._validate_payment_receipt_parameters(
            'authcode', 'order_number', 'timestamp', 'paid', 'method')

    def test_validate_payment_receipt_parameters_fail(self):
        client = Client()
        flexmock(client) \
            .should_receive('_calculate_payment_receipt_hash') \
            .with_args('order_number', 'timestamp', 'paid', 'method') \
            .and_return('not_ok')
        assert not client._validate_payment_receipt_parameters(
            'authcode', 'order_number', 'timestamp', 'paid', 'method')


verkkomaksut.py
# -*- coding: utf-8 -*-
"""
    verkkomaksut
    ~~~~~~~~~~~~

    Python wrapper for the JSON API of Suomen Verkkomaksut.

    :copyright: (c) 2012 by Janne Vanhala.
    :license: BSD, see LICENSE for more details.
"""
import hashlib
import json
import requests


class VerkkomaksutException(Exception):
    """This exception is raised when the request made to the Verkkomaksut API
    is invalid, or some other error occurs in the usage of the API."""

    def __init__(self, code, message):
        #: Error code is a unique string identifying the error. Possible error
        #: codes are listed in the `documentation`_ of Suomen Verkkomaksut
        #: REST API .
        #:
        #: .. _documentation: http://docs.verkkomaksut.fi/en/ch03s03.html
        self.code = code

        #: An error description of the error in chosen localization. This error
        #: description is not meant to be displayed to the end-user.
        self.message = message


class Contact(object):
    """This class represents the payer of a payment."""

    def __init__(self, first_name, last_name, email, street, postal_code,
                       postal_office, country, telephone=None, mobile=None,
                       company_name=None):
        #: Payer's first name.
        self.first_name = first_name

        #: Payer's surname.
        self.last_name = last_name

        #: Payer's email address.
        self.email = email

        #: Company name.
        self.company_name = company_name

        #: Payer's telephone number.
        self.telephone = telephone

        #: Payer's mobile number.
        self.mobile = mobile

        #: Payer's street address.
        self.street = street

        #: Payer's postal code.
        self.postal_code = postal_code

        #: Payer's post office.
        self.postal_office = postal_office

        #: Payer's country.  The data are sent as a two-numbered character
        #: string in ISO-3166-1 standard format.  For example, Finnish is FI
        #: and Swedish SE.  The data are used for verifying credit history,
        #: and is thus required.
        self.country = country

    @property
    def json(self):
        """JSON representation of this contact."""
        return {
            'telephone': self.telephone,
            'mobile': self.mobile,
            'email': self.email,
            'firstName': self.first_name,
            'lastName': self.last_name,
            'companyName': self.company_name,
            'address': {
                'street': self.street,
                'postalCode': self.postal_code,
                'postalOffice': self.postal_office,
                'country': self.country
            }
        }


class Payment(object):
    def __init__(self, order_number, contact, success_url, failure_url,
                       notification_url, **options):
        #: Order number is a string of characters identifying the customer's
        #: purchase and the used webshop software creates it.
        self.order_number = order_number

        #: Reference number is sent to bank by default and is automatically
        #: created.  In those payment methods that are used as an interface,
        #: this field can contain own reference number, which is sent to the
        #: bank service instead of the automatically generated reference
        #: number.
        self.reference_number = options.get('reference_number')

        #: Any data about the order in text format can be sent to the payment
        #: system.  The most usual pieces of data are customer name and contact
        #: information and order product information.  They are shown in the
        #: Merchant's Panel in payment details.
        self.description = options.get('description')

        #: Payment currency.  Value must EUR for the Finnish banks, otherwise
        #: the payment will not be accepted.
        self.currency = options.get('currency', 'EUR')

        #: Localisation defines default language for the payment method
        #: selection page and presentation format for the sums.  Available
        #: localisations are "fi_FI", "sv_SE" and "en_US". The default
        #: localisation is always "fi_FI".
        self.locale = options.get('locale', 'fi_FI')

        #: A flag indicating whether the product row prices include value
        #: added tax.  If `True` VAT is included in the shown price; otherwise
        #: it will be added.  Therefore, set this to `True`, if the prices in
        #: your webshop include value added tax, and `False` if the prices do
        #: not include value added tax.
        self.include_vat = options.get('include_vat', True)

        self.contact = contact

        #: A list of products. There must be at least one product, and the
        #: maximum number of products is 500.
        self.products = []

        #: URL to which user is directed after a successful payment.
        self.success_url = success_url

        #: URL to which user is directed after a cancelled or failed payment.
        self.failure_url = failure_url

        #: URL to which user is directed, if the payment is pending.  The
        #: status is with NetPosti payment method.  After the actual payment,
        #: the payment is signed for receipt with notify request.
        self.pending_url = options.get('pending_url')

        #: URL requested when the payment is marked as successful.  The URL is
        #: requested with the same GET parameters as success address when the
        #: payment is made.  Notification request is typically executed within
        #: a few minutes from the payment.
        self.notification_url = notification_url

    @property
    def currency(self):
        return self._currency

    @currency.setter
    def currency(self, value):
        if value != 'EUR':
            raise ValueError("Currently EUR is the only supported currency.")
        self._currency = value

    @property
    def locale(self):
        return self._locale

    @locale.setter
    def locale(self, value):
        if value not in ('fi_FI', 'sv_SE', 'en_US'):
            raise ValueError("Given locale is not supported: %r" % value)
        self._locale = value

    @property
    def json(self):
        return {
            'orderNumber': self.order_number,
            'referenceNumber': self.reference_number,
            'description': self.description,
            'currency': self.currency,
            'locale': self.locale,
            'urlSet': {
                'success': self.success_url,
                'failure': self.failure_url,
                'pending': self.pending_url,
                'notification': self.notification_url
            },
            'orderDetails': {
                'includeVat': '1' if self.include_vat else '0',
                'contact': self.contact.json,
                'products': [product.json for product in self.products]
            }
        }


class Product(object):
    TYPE_NORMAL = 1
    TYPE_POSTAGE = 2
    TYPE_PROCESSING = 3

    def __init__(self, title, price, vat, amount=1, code=None, discount=0, type=TYPE_NORMAL):
        #: Product name in free format.  The product title is shown in the
        #: Merchant's Panel and on Klarna service invoices on a product row.
        #: Product details are shown also on the payment method selection page.
        self.title = title

        #: Optional product number. Using a product number may help in
        #: aligning a correct product.
        self.code = code

        #: If an order consists of several samples of the same product, you
        #: can enter the number of products here and there won't be a need for
        #: adding each product as a separate row.  Usually this field contains
        #: value 1.
        self.amount = amount

        #: Price for one product.  If the field payment includes VAT, this is
        #: a price excluding VAT.  Otherwise, this is a price including VAT.
        #: The price can also be negative if you want to add discounts to the
        #: service.  However, the total amount of the product rows must always
        #: be bigger than 0.
        self.price = price

        #: Tax percentage for a product.  The value added tax in Finland for
        #: most products is 23.
        self.vat = vat

        #: If you have reduced the product price, you can show the discount
        #: percentage as a figure between 0 and 100 in this field.  Default
        #: discount value is 0.
        self.discount = discount

        #: A type can be specified for the product row.  `Product.TYPE_NORMAL`
        #: refers to a normal product row.  `Product.TYPE_POSTAGE` can be used
        #: for postage and `Product.TYPE_PROCESSING` for processing costs.
        #: `Product.TYPE_NORMAL` can be used for all rows, but postage and
        #: processing costs cannot be differentiated from other rows to the
        #: invoice.  Default value for the field is `Product.TYPE_NORMAL`.
        self.type = type

    @property
    def type(self):
        return self._type

    @type.setter
    def type(self, value):
        if value not in (
            Product.TYPE_NORMAL,
            Product.TYPE_POSTAGE,
            Product.TYPE_PROCESSING
        ):
            raise ValueError('Given product type not supported: %r' % value)
        self._type = value

    @property
    def json(self):
        return {
            'title': self.title,
            'code': self.code,
            'amount': self.amount,
            'price': self.price,
            'vat': self.vat,
            'discount': self.discount,
            'type': self.type
        }


class Client(object):
    SERVICE_URL = "https://payment.verkkomaksut.fi/api-payment/create"

    def __init__(self, merchant_id='13466',
                       merchant_secret='6pKF4jkv97zmqBJ3ZL8gUw5DfT2NMQ'):
        """
        Initialize the client with your own merchant id and merchant secret.

        :param merchant_id: Mercant ID is given to you by Suomen Verkkomaksut
            when you make the contract. Default is the test merchant account.
        :param merchant_secret: Merchant secret is given to you by Suoment
            Verkkomaksut. Default is the test merchant account.
        """
        self.merchant_id = merchant_id
        self.merchant_secret = merchant_secret
        self.session = requests.Session(
            auth=(merchant_id, merchant_secret),
            headers={
                'Accept': 'application/json',
                'Content-Type': 'application/json',
                'X-Verkkomaksut-Api-Version': '1'
            }
        )

    def create_payment(self, payment):
        """Creates a new payment and returns a `dict` with the following data:
        `orderNumber`

        :param payment: a `Payment` object

        """
        response = self.session.post(self.SERVICE_URL,
            data=json.dumps(payment.json)
        )

        if response.status_code != requests.codes.created:
            data = json.loads(response.content)
            raise VerkkomaksutException(
                code=data['errorCode'],
                message=data['errorMessage']
            )

        data = json.loads(response.content)
        return {
            'order_number': data['orderNumber'],
            'token': data['token'],
            'url': data['url']
        }

    def _calculate_payment_receipt_hash(self, authcode, *params):
        base = '|'.join(params + (self.merchant_secret,))
        return hashlib.md5(base).hexdigest().upper()

    def _validate_payment_receipt_parameters(self, authcode, *params):
        hash_ = self._calculate_payment_receipt_hash(self, authcode, *params)
        return authcode == hash_

    def validate_succesful_payment(self, authcode, order_number, timestamp,
                                   paid, method):
        """
        Validates parameters sent by Suomen Verkkomaksut to the success URL or
        pending URL after a successful payment. The parameters must be validated
        in order to avoid hacking attempts to confirm payment. Returns `True`
        when the parameters are valid, and `False` otherwise.

        :param authcode: A hash value calculated by payment system.
        :param order_number: The same order number that was previously sent to
            the payment system. Order number uniquely identifies each payment.
        :param timestamp: A Unix timestamp produced by Suomen Verkkomaksut used
            for calculating the hash.
        :param paid: A 10-character payment code, which is part of payment
            confirmation. In case of a pending payment, this parameter is
            always "0000000000".
        :param method: The payment method used.
        """
        return self._validate_payment_receipt_parameters(
            authcode, order_number, timestamp, paid, method
        )

    def validate_failed_payment(self, authcode, order_number, timestamp):
        """
        Validates parameters sent by Suomen Verkkomaksut to the failure URL
        after a cancelled or failed payment.

        :param authcode: A hash value calculated by payment system.
        :param order_number: The same order number that was previously sent to
            the payment system. Order number uniquely identifies each payment.
        :param timestamp: A Unix timestamp produced by Suomen Verkkomaksut used
            for calculating the hash.
        """
        return self._validate_payment_receipt_parameters(
            authcode, order_number, timestamp
        )
